{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "00bfacfa-b8a5-4b6f-b174-2fd602640ae7",
   "metadata": {},
   "source": [
    "Install some thigs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "026c37c5-3fc4-4eeb-8a1e-9a7bcf6f089f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
      "nbconvert 6.1.0 requires mistune<2,>=0.8.1, but you have mistune 2.0.5 which is incompatible.\n"
     ]
    }
   ],
   "source": [
    "%%sh --out temp_out\n",
    "pip install cwltool udocker sbpack"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e3996f18-46f3-451e-91e5-52be89d95395",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1;30mINFO\u001b[0m /opt/conda/bin/cwltool 3.1.20230527103121\n",
      "\u001b[1;30mINFO\u001b[0m Resolved 'step01.cwl' to 'file:///sbgenomics/workspace/step01.cwl'\n",
      "step01.cwl:138:7: Warning: checking item\n",
      "                  Warning:   Field 'class' contains undefined reference to\n",
      "                  'https://sevenbridges.comSaveLogs'\n",
      "step01.cwl:138:7: Warning: checking item\n",
      "                  Warning:   Field 'class' contains undefined reference to\n",
      "                  'https://sevenbridges.comSaveLogs'\n",
      "step01.cwl:454:7: Warning: checking item\n",
      "                  Warning:   Field 'class' contains undefined reference to\n",
      "                  'https://sevenbridges.comSaveLogs'\n",
      "step01.cwl:454:7: Warning: checking item\n",
      "                  Warning:   Field 'class' contains undefined reference to\n",
      "                  'https://sevenbridges.comSaveLogs'\n",
      "\u001b[1;30mINFO\u001b[0m step01.cwl:138:7: Unknown hint https://sevenbridges.comSaveLogs\n",
      "\u001b[1;30mINFO\u001b[0m step01.cwl:454:7: Unknown hint https://sevenbridges.comSaveLogs\n",
      "\u001b[1;30mINFO\u001b[0m [workflow ] start\n",
      "\u001b[1;30mINFO\u001b[0m [workflow ] starting step macs2_call_peaks\n",
      "\u001b[1;30mINFO\u001b[0m [step macs2_call_peaks] start\n",
      "\u001b[1;30mINFO\u001b[0m ['udocker', 'pull', 'images.sb.biodatacatalyst.nhlbi.nih.gov/andrewblair/cardiac-compendium:2023042401']\n",
      "Info: creating repo: /home/jovyan/.udocker\n",
      "Info: udocker command line interface 1.3.8\n",
      "Info: searching for udockertools 1.2.9\n",
      "Info: installing udockertools 1.2.9\n",
      "Info: installation of udockertools successful\n",
      "Info: downloading layer sha256:a603fa5e3b4127f210503aaa6189abf6286ee5a73deeaab460f8f33ebc6b64e2\n",
      "Info: downloading layer sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4\n",
      "Info: downloading layer sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4\n",
      "Info: downloading layer sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4\n",
      "Info: downloading layer sha256:e732803b8b3a678771339f313716f3acfc8e30d8efd5cee988ed1bba787e7595\n",
      "Info: downloading layer sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4\n",
      "Info: downloading layer sha256:af97898598424f5a977b99d13215ec4933881f08ea6c6c8036df18208dce1469\n",
      "Info: downloading layer sha256:4a42826e09b2bdd9660d9a324ce3e78c61752bcf4ce1294a6dc8b8624d42783a\n",
      "Info: downloading layer sha256:f73a961cfbce5a6db3d64d54010d57244261bf0cbfff05e028fcab40c18c9f28\n",
      "Info: downloading layer sha256:a4f214a619fb24fe2672e2e4ae04eca3b14362db330ad0d762179d518b16ffc1\n",
      "Info: downloading layer sha256:2fc9fc94e4c4e61abf36dc539c03a6c096389a027c903474c2caaa753d84e38c\n",
      "Info: downloading layer sha256:f03a7f3b22e42ebca5291dfb24c59f86a6a227f58b23572a02838d2e57b32a2e\n",
      "Info: downloading layer sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4\n",
      "Info: downloading layer sha256:1f6ac0cff2a016ca297de3ce27bb12d3b188f0dcc6090279740aa27739ea6fef\n",
      "Info: downloading layer sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4\n",
      "Info: downloading layer sha256:b4245fcd0e0b97694f386d75d6d9b61ef89e7735726ac12f0d59988aab2b3d57\n",
      "Info: downloading layer sha256:61680d9fdae5e58106a6bf8064d32755520813c526b81ae758bbeac52addf4f7\n",
      "Info: downloading layer sha256:fda22127a020fcc5efb7d4a0378b7d9a22a3338cd308b3f50cb3d5f96a0eca7e\n",
      "Info: downloading layer sha256:aef0b492a3a46a27e1d187b81c98a5c70b11b9724b361e77cc8a0eff0d396549\n",
      "Info: downloading layer sha256:302839ca5ff03970c353f4c98af459e7da3334e10b9727c0c2b406c5eb6c5146\n",
      "\u001b[1;30mINFO\u001b[0m [job macs2_call_peaks] /tmp/0b8usqud$ udocker \\\n",
      "    --quiet \\\n",
      "    run \\\n",
      "    --nobanner \\\n",
      "    --volume=/tmp/0b8usqud:/akiukT \\\n",
      "    --volume=/tmp/07rhzz1w:/tmp \\\n",
      "    --volume=/sbgenomics/project-files/organized_files/ABC-Enhancer-Gene-Prediction/example_chr22/input_data/Chromatin/wgEncodeUwDnaseK562AlnRep1.chr22.bam:/var/lib/cwl/stg8f77c300-d28d-4566-8253-b61c792e500c/wgEncodeUwDnaseK562AlnRep1.chr22.bam \\\n",
      "    --volume=/sbgenomics/project-files/organized_files/ABC-Enhancer-Gene-Prediction/example_chr22/input_data/Chromatin/wgEncodeUwDnaseK562AlnRep1.chr22.bam.bai:/var/lib/cwl/stg8f77c300-d28d-4566-8253-b61c792e500c/wgEncodeUwDnaseK562AlnRep1.chr22.bam.bai \\\n",
      "    --volume=/sbgenomics/project-files/ABC-Enhancer-Gene-Prediction/example_chr22/reference/chr22:/var/lib/cwl/stgd8276889-f000-401f-bc4c-764f8f45739f/chr22 \\\n",
      "    --workdir=/akiukT \\\n",
      "    --rm \\\n",
      "    --env=TMPDIR=/tmp \\\n",
      "    --env=HOME=/akiukT \\\n",
      "    images.sb.biodatacatalyst.nhlbi.nih.gov/andrewblair/cardiac-compendium:2023042401 \\\n",
      "    bash \\\n",
      "    call_peaks.sh \\\n",
      "    /var/lib/cwl/stg8f77c300-d28d-4566-8253-b61c792e500c/wgEncodeUwDnaseK562AlnRep1.chr22.bam \\\n",
      "    /var/lib/cwl/stgd8276889-f000-401f-bc4c-764f8f45739f/chr22\n",
      "INFO  @ Tue, 30 May 2023 22:43:58: \n",
      "# Command line: callpeak -t /var/lib/cwl/stg8f77c300-d28d-4566-8253-b61c792e500c/wgEncodeUwDnaseK562AlnRep1.chr22.bam -n wgEncodeUwDnaseK562AlnRep1.chr22.macs2 -f BAM -g hs -p .1 --call-summits --outdir ./\n",
      "# ARGUMENTS LIST:\n",
      "# name = wgEncodeUwDnaseK562AlnRep1.chr22.macs2\n",
      "# format = BAM\n",
      "# ChIP-seq file = ['/var/lib/cwl/stg8f77c300-d28d-4566-8253-b61c792e500c/wgEncodeUwDnaseK562AlnRep1.chr22.bam']\n",
      "# control file = None\n",
      "# effective genome size = 2.70e+09\n",
      "# band width = 300\n",
      "# model fold = [5, 50]\n",
      "# pvalue cutoff = 1.00e-01\n",
      "# qvalue will not be calculated and reported as -1 in the final output.\n",
      "# The maximum gap between significant sites is assigned as the read length/tag size.\n",
      "# The minimum length of peaks is assigned as the predicted fragment length \"d\".\n",
      "# Larger dataset will be scaled towards smaller dataset.\n",
      "# Range for calculating regional lambda is: 10000 bps\n",
      "# Broad region calling is off\n",
      "# Paired-End mode is off\n",
      "# Searching for subpeak summits is on\n",
      " \n",
      "INFO  @ Tue, 30 May 2023 22:43:58: #1 read tag files... \n",
      "INFO  @ Tue, 30 May 2023 22:43:58: #1 read treatment tags... \n",
      "INFO  @ Tue, 30 May 2023 22:44:00: 434301 reads have been read. \n",
      "INFO  @ Tue, 30 May 2023 22:44:00: #1 tag size is determined as 36 bps \n",
      "INFO  @ Tue, 30 May 2023 22:44:00: #1 tag size = 36.0 \n",
      "INFO  @ Tue, 30 May 2023 22:44:00: #1  total tags in treatment: 434301 \n",
      "INFO  @ Tue, 30 May 2023 22:44:00: #1 user defined the maximum tags... \n",
      "INFO  @ Tue, 30 May 2023 22:44:00: #1 filter out redundant tags at the same location and the same strand by allowing at most 1 tag(s) \n",
      "INFO  @ Tue, 30 May 2023 22:44:00: #1  tags after filtering in treatment: 330933 \n",
      "INFO  @ Tue, 30 May 2023 22:44:00: #1  Redundant rate of treatment: 0.24 \n",
      "INFO  @ Tue, 30 May 2023 22:44:00: #1 finished! \n",
      "INFO  @ Tue, 30 May 2023 22:44:00: #2 Build Peak Model... \n",
      "INFO  @ Tue, 30 May 2023 22:44:00: #2 looking for paired plus/minus strand peaks... \n",
      "INFO  @ Tue, 30 May 2023 22:44:00: #2 number of paired peaks: 2488 \n",
      "INFO  @ Tue, 30 May 2023 22:44:00: start model_add_line... \n",
      "INFO  @ Tue, 30 May 2023 22:44:00: start X-correlation... \n",
      "INFO  @ Tue, 30 May 2023 22:44:00: end of X-cor \n",
      "INFO  @ Tue, 30 May 2023 22:44:00: #2 finished! \n",
      "INFO  @ Tue, 30 May 2023 22:44:00: #2 predicted fragment length is 297 bps \n",
      "INFO  @ Tue, 30 May 2023 22:44:00: #2 alternative fragment length(s) may be 297 bps \n",
      "INFO  @ Tue, 30 May 2023 22:44:00: #2.2 Generate R script for model : ./wgEncodeUwDnaseK562AlnRep1.chr22.macs2_model.r \n",
      "INFO  @ Tue, 30 May 2023 22:44:00: #3 Call peaks... \n",
      "INFO  @ Tue, 30 May 2023 22:44:00: #3 Going to call summits inside each peak ... \n",
      "INFO  @ Tue, 30 May 2023 22:44:00: #3 Call peaks with given -log10pvalue cutoff: 1.00000 ... \n",
      "INFO  @ Tue, 30 May 2023 22:44:00: #3 Pre-compute pvalue-qvalue table... \n",
      "INFO  @ Tue, 30 May 2023 22:44:01: #3 Call peaks for each chromosome... \n",
      "INFO  @ Tue, 30 May 2023 22:44:09: #4 Write output xls file... ./wgEncodeUwDnaseK562AlnRep1.chr22.macs2_peaks.xls \n",
      "INFO  @ Tue, 30 May 2023 22:44:09: #4 Write peak in narrowPeak format file... ./wgEncodeUwDnaseK562AlnRep1.chr22.macs2_peaks.narrowPeak \n",
      "INFO  @ Tue, 30 May 2023 22:44:09: #4 Write summits bed file... ./wgEncodeUwDnaseK562AlnRep1.chr22.macs2_summits.bed \n",
      "INFO  @ Tue, 30 May 2023 22:44:09: Done! \n",
      "\u001b[1;30mINFO\u001b[0m [job macs2_call_peaks] Max memory used: 23MiB\n",
      "\u001b[1;30mERROR\u001b[0m \u001b[31m[job macs2_call_peaks] Job error:\n",
      "(\"Error collecting output for parameter 'peak_xls': step01.cwl:127:9: Did not find output file with glob pattern: ['*peak.xls'].\", {})\u001b[0m\n",
      "\u001b[1;30mWARNING\u001b[0m \u001b[33m[job macs2_call_peaks] completed permanentFail\u001b[0m\n",
      "\u001b[1;30mERROR\u001b[0m \u001b[31m[step macs2_call_peaks] Output is missing expected field https://api.sb.biodatacatalyst.nhlbi.nih.gov/v2/apps/dave/abc-development-scratch-project/abc_step01/10/raw/#macs2_call_peaks/sorted_peaks\u001b[0m\n",
      "\u001b[1;30mERROR\u001b[0m \u001b[31m[step macs2_call_peaks] Output is missing expected field https://api.sb.biodatacatalyst.nhlbi.nih.gov/v2/apps/dave/abc-development-scratch-project/abc_step01/10/raw/#macs2_call_peaks/model_r_scripts\u001b[0m\n",
      "\u001b[1;30mERROR\u001b[0m \u001b[31m[step macs2_call_peaks] Output is missing expected field https://api.sb.biodatacatalyst.nhlbi.nih.gov/v2/apps/dave/abc-development-scratch-project/abc_step01/10/raw/#macs2_call_peaks/peak_xls\u001b[0m\n",
      "\u001b[1;30mERROR\u001b[0m \u001b[31m[step macs2_call_peaks] Output is missing expected field https://api.sb.biodatacatalyst.nhlbi.nih.gov/v2/apps/dave/abc-development-scratch-project/abc_step01/10/raw/#macs2_call_peaks/summits\u001b[0m\n",
      "\u001b[1;30mWARNING\u001b[0m \u001b[33m[step macs2_call_peaks] completed permanentFail\u001b[0m\n",
      "\u001b[1;30mERROR\u001b[0m \u001b[31m[workflow ] Cannot collect workflow output: Workflow contains multiple inbound links to a single parameter but MultipleInputFeatureRequirement is not declared.\u001b[0m\n",
      "\u001b[1;30mINFO\u001b[0m [workflow ] completed permanentFail\n",
      "\u001b[1;30mWARNING\u001b[0m \u001b[33mFinal process status is permanentFail\u001b[0m\n"
     ]
    },
    {
     "ename": "CalledProcessError",
     "evalue": "Command 'b'cwltool --udocker --verbose step01.cwl \\\\\\n    --bam /sbgenomics/project-files/organized_files/ABC-Enhancer-Gene-Prediction/example_chr22/input_data/Chromatin/wgEncodeUwDnaseK562AlnRep1.chr22.bam \\\\\\n    --sort_order /sbgenomics/project-files/ABC-Enhancer-Gene-Prediction/example_chr22/reference/chr22 \\\\\\n    --regions_includelist /sbgenomics/project-files/organized_files/ABC-Enhancer-Gene-Prediction/example_chr22/reference/RefSeqCurated.170308.bed.CollapsedGeneBounds.TSS500bp.chr22.bed \\\\\\n    --regions_blocklist /sbgenomics/project-files/organized_files/ABC-Enhancer-Gene-Prediction/reference/wgEncodeHg19ConsensusSignalArtifactRegions.bed\\n'' returned non-zero exit status 1.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mCalledProcessError\u001b[0m                        Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_69/3568948680.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mget_ipython\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrun_cell_magic\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'sh'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'--out temp_out'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'cwltool --udocker --verbose step01.cwl \\\\\\n    --bam /sbgenomics/project-files/organized_files/ABC-Enhancer-Gene-Prediction/example_chr22/input_data/Chromatin/wgEncodeUwDnaseK562AlnRep1.chr22.bam \\\\\\n    --sort_order /sbgenomics/project-files/ABC-Enhancer-Gene-Prediction/example_chr22/reference/chr22 \\\\\\n    --regions_includelist /sbgenomics/project-files/organized_files/ABC-Enhancer-Gene-Prediction/example_chr22/reference/RefSeqCurated.170308.bed.CollapsedGeneBounds.TSS500bp.chr22.bed \\\\\\n    --regions_blocklist /sbgenomics/project-files/organized_files/ABC-Enhancer-Gene-Prediction/reference/wgEncodeHg19ConsensusSignalArtifactRegions.bed\\n'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/opt/conda/lib/python3.9/site-packages/IPython/core/interactiveshell.py\u001b[0m in \u001b[0;36mrun_cell_magic\u001b[0;34m(self, magic_name, line, cell)\u001b[0m\n\u001b[1;32m   2401\u001b[0m             \u001b[0;32mwith\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbuiltin_trap\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2402\u001b[0m                 \u001b[0margs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mmagic_arg_s\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcell\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2403\u001b[0;31m                 \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2404\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2405\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.9/site-packages/IPython/core/magics/script.py\u001b[0m in \u001b[0;36mnamed_script_magic\u001b[0;34m(line, cell)\u001b[0m\n\u001b[1;32m    140\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    141\u001b[0m                 \u001b[0mline\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mscript\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 142\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshebang\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mline\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcell\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    143\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    144\u001b[0m         \u001b[0;31m# write a basic docstring:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.9/site-packages/decorator.py\u001b[0m in \u001b[0;36mfun\u001b[0;34m(*args, **kw)\u001b[0m\n\u001b[1;32m    230\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mkwsyntax\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    231\u001b[0m                 \u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkw\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfix\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkw\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msig\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 232\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mcaller\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfunc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mextras\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkw\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    233\u001b[0m     \u001b[0mfun\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__name__\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__name__\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    234\u001b[0m     \u001b[0mfun\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__doc__\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__doc__\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.9/site-packages/IPython/core/magic.py\u001b[0m in \u001b[0;36m<lambda>\u001b[0;34m(f, *a, **k)\u001b[0m\n\u001b[1;32m    185\u001b[0m     \u001b[0;31m# but it's overkill for just that one bit of state.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    186\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mmagic_deco\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 187\u001b[0;31m         \u001b[0mcall\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mlambda\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    188\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    189\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mcallable\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.9/site-packages/IPython/core/magics/script.py\u001b[0m in \u001b[0;36mshebang\u001b[0;34m(self, line, cell)\u001b[0m\n\u001b[1;32m    243\u001b[0m             \u001b[0msys\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstderr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mflush\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    244\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0margs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mraise_error\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreturncode\u001b[0m\u001b[0;34m!=\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 245\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mCalledProcessError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreturncode\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcell\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0moutput\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mout\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstderr\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0merr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    246\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    247\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_run_script\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mp\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcell\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mto_close\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mCalledProcessError\u001b[0m: Command 'b'cwltool --udocker --verbose step01.cwl \\\\\\n    --bam /sbgenomics/project-files/organized_files/ABC-Enhancer-Gene-Prediction/example_chr22/input_data/Chromatin/wgEncodeUwDnaseK562AlnRep1.chr22.bam \\\\\\n    --sort_order /sbgenomics/project-files/ABC-Enhancer-Gene-Prediction/example_chr22/reference/chr22 \\\\\\n    --regions_includelist /sbgenomics/project-files/organized_files/ABC-Enhancer-Gene-Prediction/example_chr22/reference/RefSeqCurated.170308.bed.CollapsedGeneBounds.TSS500bp.chr22.bed \\\\\\n    --regions_blocklist /sbgenomics/project-files/organized_files/ABC-Enhancer-Gene-Prediction/reference/wgEncodeHg19ConsensusSignalArtifactRegions.bed\\n'' returned non-zero exit status 1."
     ]
    }
   ],
   "source": [
    "%%sh --out temp_out\n",
    "cwltool --udocker --verbose step01.cwl \\\n",
    "    --bam /sbgenomics/project-files/organized_files/ABC-Enhancer-Gene-Prediction/example_chr22/input_data/Chromatin/wgEncodeUwDnaseK562AlnRep1.chr22.bam \\\n",
    "    --sort_order /sbgenomics/project-files/ABC-Enhancer-Gene-Prediction/example_chr22/reference/chr22 \\\n",
    "    --regions_includelist /sbgenomics/project-files/organized_files/ABC-Enhancer-Gene-Prediction/example_chr22/reference/RefSeqCurated.170308.bed.CollapsedGeneBounds.TSS500bp.chr22.bed \\\n",
    "    --regions_blocklist /sbgenomics/project-files/organized_files/ABC-Enhancer-Gene-Prediction/reference/wgEncodeHg19ConsensusSignalArtifactRegions.bed\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "863d675e-5391-49cf-bc23-a4b582b157fa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: cwltool [-h] [--basedir BASEDIR] [--outdir OUTDIR] [--log-dir LOG_DIR]\n",
      "               [--parallel]\n",
      "               [--preserve-environment ENVVAR | --preserve-entire-environment]\n",
      "               [--rm-container | --leave-container]\n",
      "               [--cidfile-dir CIDFILE_DIR] [--cidfile-prefix CIDFILE_PREFIX]\n",
      "               [--tmpdir-prefix TMPDIR_PREFIX]\n",
      "               [--tmp-outdir-prefix TMP_OUTDIR_PREFIX | --cachedir CACHEDIR]\n",
      "               [--rm-tmpdir | --leave-tmpdir]\n",
      "               [--move-outputs | --leave-outputs | --copy-outputs]\n",
      "               [--enable-pull | --disable-pull]\n",
      "               [--rdf-serializer RDF_SERIALIZER] [--eval-timeout EVAL_TIMEOUT]\n",
      "               [--provenance PROVENANCE] [--enable-user-provenance]\n",
      "               [--disable-user-provenance] [--enable-host-provenance]\n",
      "               [--disable-host-provenance] [--orcid ORCID]\n",
      "               [--full-name CWL_FULL_NAME]\n",
      "               [--print-rdf | --print-dot | --print-pre | --print-deps | --print-input-deps | --pack | --version | --validate | --print-supported-versions | --print-subgraph | --print-targets | --make-template]\n",
      "               [--strict | --non-strict] [--skip-schemas]\n",
      "               [--no-doc-cache | --doc-cache] [--verbose | --quiet | --debug]\n",
      "               [--write-summary WRITE_SUMMARY] [--strict-memory-limit]\n",
      "               [--strict-cpu-limit] [--timestamps] [--js-console]\n",
      "               [--disable-js-validation]\n",
      "               [--js-hint-options-file JS_HINT_OPTIONS_FILE]\n",
      "               [--user-space-docker-cmd CMD | --udocker | --singularity | --podman | --no-container]\n",
      "               [--tool-help] [--relative-deps {primary,cwd}] [--enable-dev]\n",
      "               [--enable-ext] [--enable-color | --disable-color]\n",
      "               [--default-container DEFAULT_CONTAINER] [--no-match-user]\n",
      "               [--custom-net CUSTOM_NET]\n",
      "               [--enable-ga4gh-tool-registry | --disable-ga4gh-tool-registry]\n",
      "               [--add-ga4gh-tool-registry GA4GH_TOOL_REGISTRIES]\n",
      "               [--on-error {stop,continue}]\n",
      "               [--compute-checksum | --no-compute-checksum]\n",
      "               [--relax-path-checks] [--force-docker-pull] [--no-read-only]\n",
      "               [--overrides OVERRIDES]\n",
      "               [--target TARGET | --single-step SINGLE_STEP | --single-process SINGLE_PROCESS]\n",
      "               [--mpi-config-file MPI_CONFIG_FILE]\n",
      "               [cwl_document] ...\n",
      "\n",
      "Reference executor for Common Workflow Language standards. Not for production\n",
      "use.\n",
      "\n",
      "positional arguments:\n",
      "  cwl_document          path or URL to a CWL Workflow, CommandLineTool, or\n",
      "                        ExpressionTool. If the `inputs_object` has a\n",
      "                        `cwl:tool` field indicating the path or URL to the\n",
      "                        cwl_document, then the `cwl_document` argument is\n",
      "                        optional.\n",
      "  inputs_object         path or URL to a YAML or JSON formatted description of\n",
      "                        the required input values for the given\n",
      "                        `cwl_document`.\n",
      "\n",
      "optional arguments:\n",
      "  -h, --help            show this help message and exit\n",
      "  --basedir BASEDIR\n",
      "  --outdir OUTDIR       Output directory. The default is the current\n",
      "                        directory.\n",
      "  --log-dir LOG_DIR     Log your tools stdout/stderr to this location outside\n",
      "                        of container This will only log stdout/stderr if you\n",
      "                        specify stdout/stderr in their respective fields or\n",
      "                        capture it as an output\n",
      "  --parallel            [experimental] Run jobs in parallel.\n",
      "  --preserve-environment ENVVAR\n",
      "                        Preserve specific environment variable when running\n",
      "                        CommandLineTools. May be provided multiple times. By\n",
      "                        default PATH is preserved when not running in a\n",
      "                        container.\n",
      "  --preserve-entire-environment\n",
      "                        Preserve all environment variables when running\n",
      "                        CommandLineTools without a software container.\n",
      "  --rm-container        Delete Docker container used by jobs after they exit\n",
      "                        (default)\n",
      "  --leave-container     Do not delete Docker container used by jobs after they\n",
      "                        exit\n",
      "  --tmpdir-prefix TMPDIR_PREFIX\n",
      "                        Path prefix for temporary directories. If --tmpdir-\n",
      "                        prefix is not provided, then the prefix for temporary\n",
      "                        directories is influenced by the value of the TMPDIR,\n",
      "                        TEMP, or TMP environment variables. Taking those into\n",
      "                        consideration, the current default is /tmp/.\n",
      "  --tmp-outdir-prefix TMP_OUTDIR_PREFIX\n",
      "                        Path prefix for intermediate output directories.\n",
      "                        Defaults to the value of --tmpdir-prefix.\n",
      "  --cachedir CACHEDIR   Directory to cache intermediate workflow outputs to\n",
      "                        avoid recomputing steps. Can be very helpful in the\n",
      "                        development and troubleshooting of CWL documents.\n",
      "  --rm-tmpdir           Delete intermediate temporary directories (default)\n",
      "  --leave-tmpdir        Do not delete intermediate temporary directories\n",
      "  --move-outputs        Move output files to the workflow output directory and\n",
      "                        delete intermediate output directories (default).\n",
      "  --leave-outputs       Leave output files in intermediate output directories.\n",
      "  --copy-outputs        Copy output files to the workflow output directory and\n",
      "                        don't delete intermediate output directories.\n",
      "  --enable-pull         Try to pull Docker images\n",
      "  --disable-pull        Do not try to pull Docker images\n",
      "  --rdf-serializer RDF_SERIALIZER\n",
      "                        Output RDF serialization format used by --print-rdf\n",
      "                        (one of turtle (default), n3, nt, xml)\n",
      "  --eval-timeout EVAL_TIMEOUT\n",
      "                        Time to wait for a Javascript expression to evaluate\n",
      "                        before giving an error, default 60s.\n",
      "  --print-rdf           Print corresponding RDF graph for workflow and exit\n",
      "  --print-dot           Print workflow visualization in graphviz format and\n",
      "                        exit\n",
      "  --print-pre           Print CWL document after preprocessing.\n",
      "  --print-deps          Print CWL document dependencies.\n",
      "  --print-input-deps    Print input object document dependencies.\n",
      "  --pack                Combine components into single document and print.\n",
      "  --version             Print version and exit\n",
      "  --validate            Validate CWL document only.\n",
      "  --print-supported-versions\n",
      "                        Print supported CWL specs.\n",
      "  --print-subgraph      Print workflow subgraph that will execute. Can\n",
      "                        combined with --target or --single-step\n",
      "  --print-targets       Print targets (output parameters)\n",
      "  --make-template       Generate a template input object\n",
      "  --strict              Strict validation (unrecognized or out of place fields\n",
      "                        are error)\n",
      "  --non-strict          Lenient validation (ignore unrecognized fields)\n",
      "  --skip-schemas        Skip loading of schemas\n",
      "  --no-doc-cache        Disable disk cache for documents loaded over HTTP\n",
      "  --doc-cache           Enable disk cache for documents loaded over HTTP\n",
      "  --verbose             Default logging\n",
      "  --quiet               Only print warnings and errors.\n",
      "  --debug               Print even more logging\n",
      "  --write-summary WRITE_SUMMARY, -w WRITE_SUMMARY\n",
      "                        Path to write the final output JSON object to. Default\n",
      "                        is stdout.\n",
      "  --strict-memory-limit\n",
      "                        When running with software containers and the Docker\n",
      "                        engine, pass either the calculated memory allocation\n",
      "                        from ResourceRequirements or the default of 1 gigabyte\n",
      "                        to Docker's --memory option.\n",
      "  --strict-cpu-limit    When running with software containers and the Docker\n",
      "                        engine, pass either the calculated cpu allocation from\n",
      "                        ResourceRequirements or the default of 1 core to\n",
      "                        Docker's --cpu option. Requires docker version >=\n",
      "                        v1.13.\n",
      "  --timestamps          Add timestamps to the errors, warnings, and\n",
      "                        notifications.\n",
      "  --js-console          Enable javascript console output\n",
      "  --disable-js-validation\n",
      "                        Disable javascript validation.\n",
      "  --js-hint-options-file JS_HINT_OPTIONS_FILE\n",
      "                        File of options to pass to jshint. This includes the\n",
      "                        added option \"includewarnings\".\n",
      "  --user-space-docker-cmd CMD\n",
      "                        (Linux/OS X only) Specify the path to udocker. Implies\n",
      "                        --udocker\n",
      "  --udocker             (Linux/OS X only) Use the udocker runtime for running\n",
      "                        containers (equivalent to --user-space-docker-\n",
      "                        cmd=udocker).\n",
      "  --singularity         [experimental] Use Singularity runtime for running\n",
      "                        containers. Requires Singularity v2.6.1+ and Linux\n",
      "                        with kernel version v3.18+ or with overlayfs support\n",
      "                        backported.\n",
      "  --podman              [experimental] Use Podman runtime for running\n",
      "                        containers.\n",
      "  --no-container        Do not execute jobs in a Docker container, even when\n",
      "                        `DockerRequirement` is specified under `hints`.\n",
      "  --tool-help           Print command line help for tool\n",
      "  --relative-deps {primary,cwd}\n",
      "                        When using --print-deps, print paths relative to\n",
      "                        primary file or current working directory.\n",
      "  --enable-dev          Enable loading and running unofficial development\n",
      "                        versions of the CWL standards.\n",
      "  --enable-ext          Enable loading and running 'cwltool:' extensions to\n",
      "                        the CWL standards.\n",
      "  --enable-color        Enable logging color (default enabled)\n",
      "  --disable-color       Disable colored logging (default false)\n",
      "  --default-container DEFAULT_CONTAINER\n",
      "                        Specify a default software container to use for any\n",
      "                        CommandLineTool without a DockerRequirement.\n",
      "  --no-match-user       Disable passing the current uid to `docker run --user`\n",
      "  --custom-net CUSTOM_NET\n",
      "                        Passed to `docker run` as the '--net' parameter when\n",
      "                        NetworkAccess is true, which is its default setting.\n",
      "  --enable-ga4gh-tool-registry\n",
      "                        Enable tool resolution using GA4GH tool registry API\n",
      "  --disable-ga4gh-tool-registry\n",
      "                        Disable tool resolution using GA4GH tool registry API\n",
      "  --add-ga4gh-tool-registry GA4GH_TOOL_REGISTRIES\n",
      "                        Add a GA4GH tool registry endpoint to use for\n",
      "                        resolution, default ['https://dockstore.org/api']\n",
      "  --on-error {stop,continue}\n",
      "                        Desired workflow behavior when a step fails. One of\n",
      "                        'stop' (do not submit any more steps) or 'continue'\n",
      "                        (may submit other steps that are not downstream from\n",
      "                        the error). Default is 'stop'.\n",
      "  --compute-checksum    Compute checksum of contents while collecting outputs\n",
      "  --no-compute-checksum\n",
      "                        Do not compute checksum of contents while collecting\n",
      "                        outputs\n",
      "  --relax-path-checks   Relax requirements on path names to permit spaces and\n",
      "                        hash characters.\n",
      "  --force-docker-pull   Pull latest software container image even if it is\n",
      "                        locally present\n",
      "  --no-read-only        Do not set root directory in the container as read-\n",
      "                        only\n",
      "  --overrides OVERRIDES\n",
      "                        Read process requirement overrides from file.\n",
      "  --target TARGET, -t TARGET\n",
      "                        Only execute steps that contribute to listed targets\n",
      "                        (can be provided more than once).\n",
      "  --single-step SINGLE_STEP\n",
      "                        Only executes a single step in a workflow. The input\n",
      "                        object must match that step's inputs. Can be combined\n",
      "                        with --print-subgraph.\n",
      "  --single-process SINGLE_PROCESS\n",
      "                        Only executes the underlying Process (CommandLineTool,\n",
      "                        ExpressionTool, or sub-Workflow) for the given step in\n",
      "                        a workflow. This will not include any step-level\n",
      "                        processing: 'scatter', 'when'; and there will be no\n",
      "                        processing of step-level 'default', or 'valueFrom'\n",
      "                        input modifiers. However, requirements/hints from the\n",
      "                        step or parent workflow(s) will be inherited as\n",
      "                        usual.The input object must match that Process's\n",
      "                        inputs.\n",
      "  --mpi-config-file MPI_CONFIG_FILE\n",
      "                        Platform specific configuration for MPI (parallel\n",
      "                        launcher, its flag etc). See README section 'Running\n",
      "                        MPI-based tools' for details of the format.\n",
      "\n",
      "Options for recording the Docker container identifier into a file.:\n",
      "  --cidfile-dir CIDFILE_DIR\n",
      "                        Store the Docker container ID into a file in the\n",
      "                        specified directory.\n",
      "  --cidfile-prefix CIDFILE_PREFIX\n",
      "                        Specify a prefix to the container ID filename. Final\n",
      "                        file name will be followed by a timestamp. The default\n",
      "                        is no prefix.\n",
      "\n",
      "Options for recording provenance information of the execution:\n",
      "  --provenance PROVENANCE\n",
      "                        Save provenance to specified folder as a Research\n",
      "                        Object that captures and aggregates workflow execution\n",
      "                        and data products.\n",
      "  --enable-user-provenance\n",
      "                        Record user account info as part of provenance.\n",
      "  --disable-user-provenance\n",
      "                        Do not record user account info in provenance.\n",
      "  --enable-host-provenance\n",
      "                        Record host info as part of provenance.\n",
      "  --disable-host-provenance\n",
      "                        Do not record host info in provenance.\n",
      "  --orcid ORCID         Record user ORCID identifier as part of provenance,\n",
      "                        e.g. https://orcid.org/0000-0002-1825-0097 or\n",
      "                        0000-0002-1825-0097. Alternatively the environment\n",
      "                        variable ORCID may be set.\n",
      "  --full-name CWL_FULL_NAME\n",
      "                        Record full name of user as part of provenance, e.g.\n",
      "                        Josiah Carberry. You may need to use shell quotes to\n",
      "                        preserve spaces. Alternatively the environment\n",
      "                        variable CWL_FULL_NAME may be set.\n"
     ]
    }
   ],
   "source": [
    "%%sh\n",
    "cwltool -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b96de78-82a2-4cf8-a925-728bb1bea455",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
